File: apps/js-sdk/firecrawl/src/v2/methods/developer.ts
import type { DeveloperSearchOptions, DeveloperSearchResponse } from "../types";
import { HttpClient } from "../utils/httpClient";
import {
  normalizeAxiosError,
  throwForBadResponse,
} from "../utils/errorHandler";

const ENDPOINT = "/v2/search/developer";

/** Search the dedicated developer index with its complete filter surface. */
export async function developerSearch(
  http: HttpClient,
  query: string,
  options: DeveloperSearchOptions = {},
): Promise<DeveloperSearchResponse> {
  if (!query || !query.trim()) throw new Error("query cannot be empty");

  try {
    const response = await http.post<DeveloperSearchResponse>(ENDPOINT, {
      query,
      ...(options.k !== undefined ? { k: options.k } : {}),
      ...(options.passages !== undefined ? { passages: options.passages } : {}),
      ...(options.types !== undefined ? { types: options.types } : {}),
      ...(options.repos !== undefined ? { repos: options.repos } : {}),
      ...(options.sources !== undefined ? { sources: options.sources } : {}),
      ...(options.language !== undefined ? { language: options.language } : {}),
      ...(options.topic !== undefined ? { topic: options.topic } : {}),
      ...(options.license !== undefined ? { license: options.license } : {}),
      ...(options.minStars !== undefined
        ? { min_stars: options.minStars }
        : {}),
      ...(options.maxStars !== undefined
        ? { max_stars: options.maxStars }
        : {}),
      ...(options.archived !== undefined ? { archived: options.archived } : {}),
      ...(options.fork !== undefined ? { fork: options.fork } : {}),
      ...(options.skills !== undefined ? { skills: options.skills } : {}),
    });

    if (response.status !== 200 || !response.data?.success) {
      throwForBadResponse(response, "search developer sources");
    }
    return response.data;
  } catch (error: any) {
    if (error?.isAxiosError) {
      return normalizeAxiosError(error, "search developer sources");
    }
    throw error;
  }
}

File: apps/python-sdk/firecrawl/v2/methods/developer.py
"""Dedicated developer-search functionality for Firecrawl v2."""

from typing import List, Literal, Optional

from ..types import (
    DeveloperSearchRequest,
    DeveloperSearchResponse,
    DeveloperSearchType,
)
from ..utils import HttpClient, handle_response_error


ENDPOINT = "/v2/search/developer"


def developer_search(
    client: HttpClient,
    query: str,
    *,
    k: Optional[int] = None,
    passages: Optional[int] = None,
    types: Optional[List[DeveloperSearchType]] = None,
    repos: Optional[List[str]] = None,
    sources: Optional[List[str]] = None,
    language: Optional[str] = None,
    topic: Optional[List[str]] = None,
    license: Optional[str] = None,
    min_stars: Optional[int] = None,
    max_stars: Optional[int] = None,
    archived: Optional[bool] = None,
    fork: Optional[bool] = None,
    skills: Optional[Literal["only"]] = None,
) -> DeveloperSearchResponse:
    """Search the developer index with its complete filter surface."""
    if not query or not query.strip():
        raise ValueError("query cannot be empty")

    request = DeveloperSearchRequest(
        query=query,
        k=k,
        passages=passages,
        types=types,
        repos=repos,
        sources=sources,
        language=language,
        topic=topic,
        license=license,
        min_stars=min_stars,
        max_stars=max_stars,
        archived=archived,
        fork=fork,
        skills=skills,
    )
    response = client.post(ENDPOINT, request.model_dump(exclude_none=True))
    if response.status_code != 200:
        handle_response_error(response, "search developer sources")
    return DeveloperSearchResponse.model_validate(response.json())
